In [324]:
#import packages
import pandas as pd
from tqdm import tqdm
import pickle
import numpy as np

## 1. Import data from Spotify and Pitchfork

In [327]:
pitchfork_df = pd.read_pickle('pitchfork_df.pickle')
features_df = pd.read_pickle('song_features_df.pkl')
objects_df = pd.read_pickle('song_objects_df.pkl')

In [328]:
pitchfork_df.head()

Album           Artist Score       Genres  \
0             Petals for Armor  Hayley Williams   7.2  Pop/R&BRock   
1                      Regresa       Buscabulla   7.7   Electronic   
2          What’s New, Tomboy?    Damien Jurado   7.3         Rock   
3                  Shadow Talk       Cafe Racer   7.4         Rock   
4  It Was Good Until It Wasn’t          Kehlani   7.7      Pop/R&B   

                                             Summary  
0  The Paramore singer’s debut solo album is emot...  
1  The Puerto Rican duo moved back to the island ...  
2  Channeling stripped-back pop songs into hushed...  
3  The Chicago band’s third album brings wiry new...  
4  The cloudy grooves of the Oakland singer’s sec...

In [329]:
features_df.head()

danceability  energy   key  loudness  mode  speechiness  acousticness  \
0         0.790   0.599     7    -7.417     1       0.0431         0.218   
1         0.570   0.339     2   -11.857     0       0.0449         0.569   
2           NaN     NaN  None       NaN  None          NaN           NaN   
3         0.877   0.598     6    -6.292     0       0.0296         0.353   
4         0.564   0.391     7    -7.816     1       0.0785         0.235   

   instrumentalness  liveness  valence    tempo time_signature  \
0          0.002570     0.118    0.362  109.978              4   
1          0.451000     0.126    0.442   88.963              4   
2               NaN       NaN      NaN      NaN           None   
3          0.001320     0.380    0.671   97.012              4   
4          0.000054     0.110    0.419  101.449              4   

                  song_id  
0  2moHnkHTSXBe9KjvPSQJvg  
1  5ccn0VZg8HQZsZXxGx7UZQ  
2                    None  
3  5ZD0jxyhmdNZ3DwDRXqV5U  
4  1y2GDXFvtHPbCkHFYg49Gn

In [330]:
objects_df.head()

name duration_ms popularity num_markets             album  \
0          Simmer      266115         64          79  Petals For Armor   
1  Leave It Alone      245076         61          79  Petals For Armor   
2        Cinnamon      211834         61          79  Petals For Armor   
3        Creepin'      178398         60          79  Petals For Armor   
4   Sudden Desire      187969         60          79  Petals For Armor   

  disc_number is_explicit track_number release_date           artist  \
0           1        True            1   2020-05-08  Hayley Williams   
1           1       False            2   2020-05-08  Hayley Williams   
2           1       False            3   2020-05-08  Hayley Williams   
3           1       False            4   2020-05-08  Hayley Williams   
4           1       False            5   2020-05-08  Hayley Williams   

                  song_id                album_id  
0  2moHnkHTSXBe9KjvPSQJvg  4HXpQ5KQBVWN25ltjnX7xa  
1  5ccn0VZg8HQZsZXxGx7UZQ  4HXpQ5KQBVWN25ltjnX7xa  
2  48pL4zB6KXWPvd7Ln33ENy  4HXpQ5KQBVWN25ltjnX7xa  
3  5ZD0jxyhmdNZ3DwDRXqV5U  4HXpQ5KQBVWN25ltjnX7xa  
4  1y2GDXFvtHPbCkHFYg49Gn  4HXpQ5KQBVWN25ltjnX7xa

## 2. Merge Spotify objects and features dataframes

In [331]:
df_spotify = pd.merge(features_df, objects_df, on='song_id')

In [332]:
#in order to see all the columns
pd.set_option('display.max_columns', None)

In [333]:
df_spotify.head()

danceability  energy   key  loudness  mode  speechiness  acousticness  \
0          0.79   0.599     7    -7.417     1       0.0431         0.218   
1          0.57   0.339     2   -11.857     0       0.0449         0.569   
2           NaN     NaN  None       NaN  None          NaN           NaN   
3           NaN     NaN  None       NaN  None          NaN           NaN   
4           NaN     NaN  None       NaN  None          NaN           NaN   

   instrumentalness  liveness  valence    tempo time_signature  \
0           0.00257     0.118    0.362  109.978              4   
1           0.45100     0.126    0.442   88.963              4   
2               NaN       NaN      NaN      NaN           None   
3               NaN       NaN      NaN      NaN           None   
4               NaN       NaN      NaN      NaN           None   

                  song_id            name duration_ms popularity num_markets  \
0  2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64          79   
1  5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61          79   
2                    None            None        None       None        None   
3                    None            None        None       None        None   
4                    None            None        None       None        None   

              album disc_number is_explicit track_number release_date  \
0  Petals For Armor           1        True            1   2020-05-08   
1  Petals For Armor           1       False            2   2020-05-08   
2              None        None        None         None         None   
3              None        None        None         None         None   
4              None        None        None         None         None   

            artist                album_id  
0  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
1  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
2             None                    None  
3             None                    None  
4             None                    None

In [334]:
df_spotify.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 855902 entries, 0 to 855901
Data columns (total 24 columns):
danceability        23738 non-null float64
energy              23738 non-null float64
key                 23738 non-null object
loudness            23738 non-null float64
mode                23738 non-null object
speechiness         23738 non-null float64
acousticness        23738 non-null float64
instrumentalness    23738 non-null float64
liveness            23738 non-null float64
valence             23738 non-null float64
tempo               23738 non-null float64
time_signature      23738 non-null object
song_id             23738 non-null object
name                23738 non-null object
duration_ms         23738 non-null object
popularity          23738 non-null object
num_markets         23738 non-null object
album               23738 non-null object
disc_number         23738 non-null object
is_explicit         23738 non-null object
track_number        23738 non-null objec

In [335]:
#remove albums with no information
df_spotify = df_spotify[df_spotify['energy'].notna()]

In [336]:
df_spotify.head()

danceability  energy key  loudness mode  speechiness  acousticness  \
0              0.790   0.599   7    -7.417    1       0.0431         0.218   
1              0.570   0.339   2   -11.857    0       0.0449         0.569   
832166         0.877   0.598   6    -6.292    0       0.0296         0.353   
832167         0.564   0.391   7    -7.816    1       0.0785         0.235   
832168         0.737   0.787   7    -4.587    1       0.0525         0.121   

        instrumentalness  liveness  valence    tempo time_signature  \
0               0.002570     0.118    0.362  109.978              4   
1               0.451000     0.126    0.442   88.963              4   
832166          0.001320     0.380    0.671   97.012              4   
832167          0.000054     0.110    0.419  101.449              4   
832168          0.000002     0.114    0.590   90.014              4   

                       song_id            name duration_ms popularity  \
0       2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64   
1       5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61   
832166  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'      178398         60   
832167  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire      187969         60   
832168  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse      199412         64   

       num_markets             album disc_number is_explicit track_number  \
0               79  Petals For Armor           1        True            1   
1               79  Petals For Armor           1       False            2   
832166          79  Petals For Armor           1       False            4   
832167          79  Petals For Armor           1       False            5   
832168          79  Petals For Armor           2       False            1   

       release_date           artist                album_id  
0        2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
1        2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832166   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832167   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa  
832168   2020-05-08  Hayley Williams  4HXpQ5KQBVWN25ltjnX7xa

In [337]:
df_spotify.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 23738 entries, 0 to 855901
Data columns (total 24 columns):
danceability        23738 non-null float64
energy              23738 non-null float64
key                 23738 non-null object
loudness            23738 non-null float64
mode                23738 non-null object
speechiness         23738 non-null float64
acousticness        23738 non-null float64
instrumentalness    23738 non-null float64
liveness            23738 non-null float64
valence             23738 non-null float64
tempo               23738 non-null float64
time_signature      23738 non-null object
song_id             23738 non-null object
name                23738 non-null object
duration_ms         23738 non-null object
popularity          23738 non-null object
num_markets         23738 non-null object
album               23738 non-null object
disc_number         23738 non-null object
is_explicit         23738 non-null object
track_number        23738 non-null object

## 3. Merge Spotify and Pitchfork info

In [339]:
#apply lower case to albums and titles for merging
pitchfork_df['Album'] = pitchfork_df['Album'].astype(str).apply(lambda x: x.lower())
pitchfork_df['Artist'] = pitchfork_df['Artist'].astype(str).apply(lambda x: x.lower())

In [340]:
pitchfork_df.head()

Album           Artist Score       Genres  \
0             petals for armor  hayley williams   7.2  Pop/R&BRock   
1                      regresa       buscabulla   7.7   Electronic   
2          what’s new, tomboy?    damien jurado   7.3         Rock   
3                  shadow talk       cafe racer   7.4         Rock   
4  it was good until it wasn’t          kehlani   7.7      Pop/R&B   

                                             Summary  
0  The Paramore singer’s debut solo album is emot...  
1  The Puerto Rican duo moved back to the island ...  
2  Channeling stripped-back pop songs into hushed...  
3  The Chicago band’s third album brings wiry new...  
4  The cloudy grooves of the Oakland singer’s sec...

In [341]:
#rename columns to match for merging
df_spotify.rename(columns={"album": "Album", "artist": "Artist"}, inplace=True)

In [342]:
#apply lower case to albums and titles for merging
df_spotify['Album'] = df_spotify['Album'].astype(str).apply(lambda x: x.lower())
df_spotify['Artist'] = df_spotify['Artist'].astype(str).apply(lambda x: x.lower())

In [343]:
df_spotify.head()

danceability  energy key  loudness mode  speechiness  acousticness  \
0              0.790   0.599   7    -7.417    1       0.0431         0.218   
1              0.570   0.339   2   -11.857    0       0.0449         0.569   
832166         0.877   0.598   6    -6.292    0       0.0296         0.353   
832167         0.564   0.391   7    -7.816    1       0.0785         0.235   
832168         0.737   0.787   7    -4.587    1       0.0525         0.121   

        instrumentalness  liveness  valence    tempo time_signature  \
0               0.002570     0.118    0.362  109.978              4   
1               0.451000     0.126    0.442   88.963              4   
832166          0.001320     0.380    0.671   97.012              4   
832167          0.000054     0.110    0.419  101.449              4   
832168          0.000002     0.114    0.590   90.014              4   

                       song_id            name duration_ms popularity  \
0       2moHnkHTSXBe9KjvPSQJvg          Simmer      266115         64   
1       5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone      245076         61   
832166  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'      178398         60   
832167  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire      187969         60   
832168  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse      199412         64   

       num_markets             Album disc_number is_explicit track_number  \
0               79  petals for armor           1        True            1   
1               79  petals for armor           1       False            2   
832166          79  petals for armor           1       False            4   
832167          79  petals for armor           1       False            5   
832168          79  petals for armor           2       False            1   

       release_date           Artist                album_id  
0        2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
1        2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832166   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832167   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa  
832168   2020-05-08  hayley williams  4HXpQ5KQBVWN25ltjnX7xa

In [344]:
#merge spotify and pitchfork dataframes
df = pitchfork_df.merge(df_spotify, how='inner')

In [345]:
#dropping rows with value == null
df = df[(df != 'null').all(axis=1)]

/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


In [346]:
df.head()

Album           Artist Score       Genres  \
0  petals for armor  hayley williams   7.2  Pop/R&BRock   
1  petals for armor  hayley williams   7.2  Pop/R&BRock   
2  petals for armor  hayley williams   7.2  Pop/R&BRock   
3  petals for armor  hayley williams   7.2  Pop/R&BRock   
4  petals for armor  hayley williams   7.2  Pop/R&BRock   

                                             Summary  danceability  energy  \
0  The Paramore singer’s debut solo album is emot...         0.790   0.599   
1  The Paramore singer’s debut solo album is emot...         0.570   0.339   
2  The Paramore singer’s debut solo album is emot...         0.877   0.598   
3  The Paramore singer’s debut solo album is emot...         0.564   0.391   
4  The Paramore singer’s debut solo album is emot...         0.737   0.787   

  key  loudness mode  speechiness  acousticness  instrumentalness  liveness  \
0   7    -7.417    1       0.0431         0.218          0.002570     0.118   
1   2   -11.857    0       0.0449         0.569          0.451000     0.126   
2   6    -6.292    0       0.0296         0.353          0.001320     0.380   
3   7    -7.816    1       0.0785         0.235          0.000054     0.110   
4   7    -4.587    1       0.0525         0.121          0.000002     0.114   

   valence    tempo time_signature                 song_id            name  \
0    0.362  109.978              4  2moHnkHTSXBe9KjvPSQJvg          Simmer   
1    0.442   88.963              4  5ccn0VZg8HQZsZXxGx7UZQ  Leave It Alone   
2    0.671   97.012              4  5ZD0jxyhmdNZ3DwDRXqV5U        Creepin'   
3    0.419  101.449              4  1y2GDXFvtHPbCkHFYg49Gn   Sudden Desire   
4    0.590   90.014              4  5RdkyZw1KQ1v0EbquUAsx2      Dead Horse   

  duration_ms popularity num_markets disc_number is_explicit track_number  \
0      266115         64          79           1        True            1   
1      245076         61          79           1       False            2   
2      178398         60          79           1       False            4   
3      187969         60          79           1       False            5   
4      199412         64          79           2       False            1   

  release_date                album_id  
0   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
1   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
2   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
3   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa  
4   2020-05-08  4HXpQ5KQBVWN25ltjnX7xa

In [347]:
df['album_id'].nunique()

1617

#### I now have 1,617 unique albums to work with

## 4. Create dataframe for modeling

In [348]:
#isolate desired columns
model_df = df[['Score','Genres', 'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness',
               'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature',
               'song_id', 'duration_ms',  'popularity', 'num_markets', 'disc_number', 'is_explicit',
               'track_number', 'album_id']]

In [349]:
model_df.head()

Score       Genres  danceability  energy key  loudness mode  speechiness  \
0   7.2  Pop/R&BRock         0.790   0.599   7    -7.417    1       0.0431   
1   7.2  Pop/R&BRock         0.570   0.339   2   -11.857    0       0.0449   
2   7.2  Pop/R&BRock         0.877   0.598   6    -6.292    0       0.0296   
3   7.2  Pop/R&BRock         0.564   0.391   7    -7.816    1       0.0785   
4   7.2  Pop/R&BRock         0.737   0.787   7    -4.587    1       0.0525   

   acousticness  instrumentalness  liveness  valence    tempo time_signature  \
0         0.218          0.002570     0.118    0.362  109.978              4   
1         0.569          0.451000     0.126    0.442   88.963              4   
2         0.353          0.001320     0.380    0.671   97.012              4   
3         0.235          0.000054     0.110    0.419  101.449              4   
4         0.121          0.000002     0.114    0.590   90.014              4   

                  song_id duration_ms popularity num_markets disc_number  \
0  2moHnkHTSXBe9KjvPSQJvg      266115         64          79           1   
1  5ccn0VZg8HQZsZXxGx7UZQ      245076         61          79           1   
2  5ZD0jxyhmdNZ3DwDRXqV5U      178398         60          79           1   
3  1y2GDXFvtHPbCkHFYg49Gn      187969         60          79           1   
4  5RdkyZw1KQ1v0EbquUAsx2      199412         64          79           2   

  is_explicit track_number                album_id  
0        True            1  4HXpQ5KQBVWN25ltjnX7xa  
1       False            2  4HXpQ5KQBVWN25ltjnX7xa  
2       False            4  4HXpQ5KQBVWN25ltjnX7xa  
3       False            5  4HXpQ5KQBVWN25ltjnX7xa  
4       False            1  4HXpQ5KQBVWN25ltjnX7xa

In [350]:
model_df.Genres.value_counts()

Rock                                  6040
Rap                                   3935
Pop/R&B                               2446
Electronic                            2312
Experimental                          1448
Folk/Country                           850
Jazz                                   469
Metal                                  442
ElectronicRock                         375
ExperimentalRock                       104
Global                                  96
Pop/R&BRock                             96
JazzPop/R&B                             67
Pop/R&BRap                              60
ElectronicPop/R&B                       44
ElectronicExperimental                  42
RapRock                                 31
ExperimentalElectronic                  29
GlobalPop/R&B                           27
ExperimentalRap                         25
Folk/CountryPop/R&B                     23
MetalRock                               21
RockExperimental                        16
Pop/R&BElec

In [351]:
#replace subgenres with main genre
model_df['Genres'].replace(['Pop/R&BRock'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RockExperimental'], 'ExperimentalRock', inplace=True)
model_df['Genres'].replace(['Pop/R&BJazz'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RapFolk/Country'], 'Rap', inplace=True)
model_df['Genres'].replace(['ElectronicJazzRock'], 'ElectronicRock', inplace=True)
model_df['Genres'].replace(['ExperimentalPop/R&B'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['Pop/R&BFolk/Country'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['RockFolk/Country'], 'Rock', inplace=True)
model_df['Genres'].replace(['Pop/R&BExperimental'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalFolk/Country'], 'Electronic', inplace=True)
model_df['Genres'].replace(['ElectronicRap'], 'Rap', inplace=True)
model_df['Genres'].replace(['ElectronicJazzPop/R&B'], 'Electronic', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalMetal'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['RockMetal'], 'Metal', inplace=True)
model_df['Genres'].replace(['RockRap'], 'Rock', inplace=True)
model_df['Genres'].replace(['ElectronicExperimentalRock'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['Pop/R&BElectronic'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['ElectronicExperimental'], 'ExperimentalElectronic', inplace=True)
model_df['Genres'].replace(['RapRock'], 'Rap', inplace=True)
model_df['Genres'].replace(['ExperimentalRap'], 'Rap', inplace=True)
model_df['Genres'].replace(['Folk/CountryPop/R&B'], 'Folk/Country', inplace=True)
model_df['Genres'].replace(['GlobalPop/R&B'], 'Global', inplace=True)
model_df['Genres'].replace(['MetalRock'], 'Metal', inplace=True)
model_df['Genres'].replace(['Pop/R&BRap'], 'Pop/R&B', inplace=True)
model_df['Genres'].replace(['JazzPop/R&B'], 'Jazz', inplace=True)
model_df['Genres'].replace(['ElectronicPop/R&B'], 'Pop/R&B', inplace=True)


/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py:6786: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [352]:
model_df.Genres.value_counts()

Rock                      6062
Rap                       4010
Pop/R&B                   2699
Electronic                2333
Experimental              1448
Folk/Country               873
Jazz                       536
Metal                      474
ElectronicRock             385
Global                     123
ExperimentalRock           120
ExperimentalElectronic      94
Name: Genres, dtype: int64

In [353]:
model_df.Genres.nunique()

12

In [367]:
#change is_explicit column bool to numeric
model_df['is_explicit'].replace(to_replace='True', value = 1, inplace=True)
model_df['is_explicit'].replace(to_replace='False', value = 0, inplace=True)

In [368]:
model_df.is_explicit.head()

0     True
1    False
2    False
3    False
4    False
Name: is_explicit, dtype: bool

In [356]:
type(model_df.Score[0])

str

In [357]:
#change score column from string to float
model_df['Score'] = model_df['Score'].astype(float)
model_df['popularity'] = model_df['popularity'].astype(float)
model_df['mode'] = model_df['mode'].astype(float)
model_df['key'] = model_df['key'].astype(float)

/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/brendonius/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataF

In [358]:
model_df.head()

Score   Genres  danceability  energy  key  loudness  mode  speechiness  \
0    7.2  Pop/R&B         0.790   0.599  7.0    -7.417   1.0       0.0431   
1    7.2  Pop/R&B         0.570   0.339  2.0   -11.857   0.0       0.0449   
2    7.2  Pop/R&B         0.877   0.598  6.0    -6.292   0.0       0.0296   
3    7.2  Pop/R&B         0.564   0.391  7.0    -7.816   1.0       0.0785   
4    7.2  Pop/R&B         0.737   0.787  7.0    -4.587   1.0       0.0525   

   acousticness  instrumentalness  liveness  valence    tempo time_signature  \
0         0.218          0.002570     0.118    0.362  109.978              4   
1         0.569          0.451000     0.126    0.442   88.963              4   
2         0.353          0.001320     0.380    0.671   97.012              4   
3         0.235          0.000054     0.110    0.419  101.449              4   
4         0.121          0.000002     0.114    0.590   90.014              4   

                  song_id duration_ms  popularity num_markets disc_number  \
0  2moHnkHTSXBe9KjvPSQJvg      266115        64.0          79           1   
1  5ccn0VZg8HQZsZXxGx7UZQ      245076        61.0          79           1   
2  5ZD0jxyhmdNZ3DwDRXqV5U      178398        60.0          79           1   
3  1y2GDXFvtHPbCkHFYg49Gn      187969        60.0          79           1   
4  5RdkyZw1KQ1v0EbquUAsx2      199412        64.0          79           2   

   is_explicit track_number                album_id  
0         True            1  4HXpQ5KQBVWN25ltjnX7xa  
1        False            2  4HXpQ5KQBVWN25ltjnX7xa  
2        False            4  4HXpQ5KQBVWN25ltjnX7xa  
3        False            5  4HXpQ5KQBVWN25ltjnX7xa  
4        False            1  4HXpQ5KQBVWN25ltjnX7xa

In [359]:
#group songs by album_id
grouped_albums = model_df.groupby('album_id')

In [360]:
grouped_albums.size()

album_id
007DWn799UWvfY1wwZeENR    15
00ojDEHvnEOPOjhHrUrKy5    10
00oz3t7cI3WfwS2oEIZD6x     9
00uqjqwLruO346z3cNlNqG    13
01ey0ETPITd3jiOiztRXzK    16
                          ..
7z4GhRfLqfSkqrj5F3Yt2B    12
7zIABwRfxyMXWDEaLm89HG    14
7zQuYvRVyQxgIlxhcSLTdW    18
7zeCZY6rQRufc8IHGKyXGX    11
7zgo8oFUtR2cVQbhDZ0o6V     6
Length: 1617, dtype: int64

In [361]:
#get mean values of numerical variables
grouped_mean_df = grouped_albums[
    'danceability', 'acousticness', 'instrumentalness', 'loudness',
    'tempo'
].agg([np.mean])


In [362]:
grouped_mean_df.columns = ["_".join(x) for x in grouped_mean_df.columns.ravel()]

In [363]:
grouped_mean_df.head()

danceability_mean  acousticness_mean  \
album_id                                                       
007DWn799UWvfY1wwZeENR           0.871667           0.046364   
00ojDEHvnEOPOjhHrUrKy5           0.724300           0.374767   
00oz3t7cI3WfwS2oEIZD6x           0.443889           0.957778   
00uqjqwLruO346z3cNlNqG           0.562154           0.464462   
01ey0ETPITd3jiOiztRXzK           0.377625           0.406806   

                        instrumentalness_mean  loudness_mean  tempo_mean  
album_id                                                                  
007DWn799UWvfY1wwZeENR               0.000119      -7.147600  126.265133  
00ojDEHvnEOPOjhHrUrKy5               0.532800     -15.015700  100.700800  
00oz3t7cI3WfwS2oEIZD6x               0.205918     -13.239333  118.242111  
00uqjqwLruO346z3cNlNqG               0.011059      -6.111769  124.775154  
01ey0ETPITd3jiOiztRXzK               0.049877     -12.121375  105.696000

In [364]:
#get max value of other numerical variables
grouped_max_df = grouped_albums[
    'energy', 'liveness', 'speechiness', 'valence', 'tempo', 'time_signature', 'popularity', 
    'Score', 'num_markets', 'is_explicit', 'mode', 'key'
].agg([np.max])


In [365]:
grouped_max_df.columns = ["_".join(x) for x in grouped_max_df.columns.ravel()]

In [366]:
grouped_max_df.head()

energy_amax  liveness_amax  speechiness_amax  \
album_id                                                               
007DWn799UWvfY1wwZeENR        0.695          0.454            0.5510   
00ojDEHvnEOPOjhHrUrKy5        0.697          0.896            0.1460   
00oz3t7cI3WfwS2oEIZD6x        0.268          0.161            0.0512   
00uqjqwLruO346z3cNlNqG        0.911          0.570            0.2120   
01ey0ETPITd3jiOiztRXzK        0.697          0.355            0.0607   

                        valence_amax  tempo_amax  time_signature_amax  \
album_id                                                                
007DWn799UWvfY1wwZeENR         0.963     173.368                    4   
00ojDEHvnEOPOjhHrUrKy5         0.874     120.006                    4   
00oz3t7cI3WfwS2oEIZD6x         0.342     207.893                    5   
00uqjqwLruO346z3cNlNqG         0.782     203.911                    4   
01ey0ETPITd3jiOiztRXzK         0.543     181.821                    4   

                        popularity_amax  Score_amax  num_markets_amax  \
album_id                                                                
007DWn799UWvfY1wwZeENR             80.0         7.8                79   
00ojDEHvnEOPOjhHrUrKy5              7.0         7.4                79   
00oz3t7cI3WfwS2oEIZD6x             45.0         8.4                 1   
00uqjqwLruO346z3cNlNqG             66.0         7.4                79   
01ey0ETPITd3jiOiztRXzK              4.0         7.1                79   

                        is_explicit_amax  mode_amax  key_amax  
album_id                                                       
007DWn799UWvfY1wwZeENR              True        1.0      10.0  
00ojDEHvnEOPOjhHrUrKy5             False        1.0      11.0  
00oz3t7cI3WfwS2oEIZD6x             False        1.0      11.0  
00uqjqwLruO346z3cNlNqG             False        1.0      11.0  
01ey0ETPITd3jiOiztRXzK              True        1.0       9.0

In [377]:
#get album length in time and number of tracks in an album
album_std_duration_df = grouped_albums.agg(album_run_time_ms = pd.NamedAgg(column='duration_ms', aggfunc=np.sum),
                                        number_of_tracks = pd.NamedAgg(column='track_number', aggfunc='count'),
                                       std_in_energy = pd.NamedAgg(column='energy', aggfunc=np.std),
                                       std_in_speechiness = pd.NamedAgg(column='speechiness', aggfunc=np.std))

In [379]:
album_std_durations_df = album_durations_df[album_durations_df.number_of_tracks <= 35]

In [380]:
album_std_durations_df.head()

album_run_time_ms  number_of_tracks  std_in_energy  \
album_id                                                                     
007DWn799UWvfY1wwZeENR            3065169                15       0.072424   
00ojDEHvnEOPOjhHrUrKy5            2886000                10       0.185253   
00oz3t7cI3WfwS2oEIZD6x            1670290                 9       0.073569   
00uqjqwLruO346z3cNlNqG            2337540                13       0.204683   
01ey0ETPITd3jiOiztRXzK            2840581                16       0.141410   

                        std_in_speechiness  
album_id                                    
007DWn799UWvfY1wwZeENR            0.162290  
00ojDEHvnEOPOjhHrUrKy5            0.041812  
00oz3t7cI3WfwS2oEIZD6x            0.007601  
00uqjqwLruO346z3cNlNqG            0.057550  
01ey0ETPITd3jiOiztRXzK            0.008499

In [381]:
#concatenate the two dataframes
album_df = pd.concat([album_std_durations_df, grouped_mean_df, grouped_max_df], axis=1, sort=False)

In [382]:
#add genres column to dataframe
genres = grouped_albums['Genres'].first()
album_df['genres'] = genres

In [383]:
album_df.genres.value_counts()

Rock                      531
Rap                       268
Electronic                219
Pop/R&B                   213
Experimental              152
Folk/Country               82
Jazz                       49
Metal                      48
ElectronicRock             22
ExperimentalRock           12
Global                     11
ExperimentalElectronic     10
Name: genres, dtype: int64

In [385]:
album_df.head()

album_run_time_ms  number_of_tracks  std_in_energy  \
007DWn799UWvfY1wwZeENR          3065169.0              15.0       0.072424   
00ojDEHvnEOPOjhHrUrKy5          2886000.0              10.0       0.185253   
00oz3t7cI3WfwS2oEIZD6x          1670290.0               9.0       0.073569   
00uqjqwLruO346z3cNlNqG          2337540.0              13.0       0.204683   
01ey0ETPITd3jiOiztRXzK          2840581.0              16.0       0.141410   

                        std_in_speechiness  danceability_mean  \
007DWn799UWvfY1wwZeENR            0.162290           0.871667   
00ojDEHvnEOPOjhHrUrKy5            0.041812           0.724300   
00oz3t7cI3WfwS2oEIZD6x            0.007601           0.443889   
00uqjqwLruO346z3cNlNqG            0.057550           0.562154   
01ey0ETPITd3jiOiztRXzK            0.008499           0.377625   

                        acousticness_mean  instrumentalness_mean  \
007DWn799UWvfY1wwZeENR           0.046364               0.000119   
00ojDEHvnEOPOjhHrUrKy5           0.374767               0.532800   
00oz3t7cI3WfwS2oEIZD6x           0.957778               0.205918   
00uqjqwLruO346z3cNlNqG           0.464462               0.011059   
01ey0ETPITd3jiOiztRXzK           0.406806               0.049877   

                        loudness_mean  tempo_mean  energy_amax  liveness_amax  \
007DWn799UWvfY1wwZeENR      -7.147600  126.265133        0.695          0.454   
00ojDEHvnEOPOjhHrUrKy5     -15.015700  100.700800        0.697          0.896   
00oz3t7cI3WfwS2oEIZD6x     -13.239333  118.242111        0.268          0.161   
00uqjqwLruO346z3cNlNqG      -6.111769  124.775154        0.911          0.570   
01ey0ETPITd3jiOiztRXzK     -12.121375  105.696000        0.697          0.355   

                        speechiness_amax  valence_amax  tempo_amax  \
007DWn799UWvfY1wwZeENR            0.5510         0.963     173.368   
00ojDEHvnEOPOjhHrUrKy5            0.1460         0.874     120.006   
00oz3t7cI3WfwS2oEIZD6x            0.0512         0.342     207.893   
00uqjqwLruO346z3cNlNqG            0.2120         0.782     203.911   
01ey0ETPITd3jiOiztRXzK            0.0607         0.543     181.821   

                        time_signature_amax  popularity_amax  Score_amax  \
007DWn799UWvfY1wwZeENR                    4             80.0         7.8   
00ojDEHvnEOPOjhHrUrKy5                    4              7.0         7.4   
00oz3t7cI3WfwS2oEIZD6x                    5             45.0         8.4   
00uqjqwLruO346z3cNlNqG                    4             66.0         7.4   
01ey0ETPITd3jiOiztRXzK                    4              4.0         7.1   

                        num_markets_amax  is_explicit_amax  mode_amax  \
007DWn799UWvfY1wwZeENR                79              True        1.0   
00ojDEHvnEOPOjhHrUrKy5                79             False        1.0   
00oz3t7cI3WfwS2oEIZD6x                 1             False        1.0   
00uqjqwLruO346z3cNlNqG                79             False        1.0   
01ey0ETPITd3jiOiztRXzK                79              True        1.0   

                        key_amax          genres  
007DWn799UWvfY1wwZeENR      10.0             Rap  
00ojDEHvnEOPOjhHrUrKy5      11.0      Electronic  
00oz3t7cI3WfwS2oEIZD6x      11.0    Folk/Country  
00uqjqwLruO346z3cNlNqG      11.0         Pop/R&B  
01ey0ETPITd3jiOiztRXzK       9.0  ElectronicRock

In [386]:
album_df = album_df[album_df['number_of_tracks'].notna()]

In [387]:
album_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1604 entries, 007DWn799UWvfY1wwZeENR to 7zgo8oFUtR2cVQbhDZ0o6V
Data columns (total 22 columns):
album_run_time_ms        1604 non-null float64
number_of_tracks         1604 non-null float64
std_in_energy            1600 non-null float64
std_in_speechiness       1600 non-null float64
danceability_mean        1604 non-null float64
acousticness_mean        1604 non-null float64
instrumentalness_mean    1604 non-null float64
loudness_mean            1604 non-null float64
tempo_mean               1604 non-null float64
energy_amax              1604 non-null float64
liveness_amax            1604 non-null float64
speechiness_amax         1604 non-null float64
valence_amax             1604 non-null float64
tempo_amax               1604 non-null float64
time_signature_amax      1604 non-null int64
popularity_amax          1604 non-null float64
Score_amax               1604 non-null float64
num_markets_amax         1604 non-null int64
is_explicit_amax 

In [388]:
#save df for modeling
with open('album_df.pkl', 'wb') as f:
    pickle.dump(album_df, f)